### PR TITLE
Updates to working with broken code

### LINKED ISSUE #73: Add in guard rails for auto-fixing code
Currently `cargo fix` will blindly apply fixes that it feels confident are the only way to fix something. This, with bugs, can cause code to stop compiling, however.

We should check that if a `cargo check` succeeds *before* rustfix runs and then fails afterwards, all fixes are backed out and a large warning/error is presented to the user. Ideally this message we present even includes information like how to report a bug to the rustfix repository and what information to include.

### LINKED ISSUE #80: Should rustfix fix broken code?
Should `cargo fix` keep going if the first `cargo check` fails? That is, if we're looking at broken code, should `cargo fix` attempt to keep going and fixing code?

In situations like this we can't actually detect if our changes caused further breakage, so we'd run the risk of making the situation *worse* by adding more errors.

### PR DESCRIPTION
**Pull request by alexcrichton**
* Don't attempt to run rustfix over already broken code
* If we generate broken code, back out all automatic changes

Closes #73
Closes #80

### CODE REVIEW

File: cargo-fix/tests/all/broken_build.rs [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+
+    // Build our rustc shim
+    p.expect_cmd("cargo build")
+        .cwd("foo")
+        .run();
+
+    // Attempt to fix code, but our shim will always fail the second compile
+    p.expect_cmd("cargo-fix fix")
+        .cwd("bar")
+        .env("RUSTC", p.root.join("foo/target/debug/foo"))
```
[the remark this answers is not in the document]
Comment (killercup) [reply cargo-fix/tests/all/broken_build.rs]: That being said, it test doesn't seem to work on Windows?

File: cargo-fix/src/main.rs [hunk 2]
```diff
@@ -91,21 +91,50 @@ fn cargo_fix_rustc() -> Result<(), Error> {
...
+        // any. If stderr is empty then there's no need for the final exec at
+        // the end, we just bail out here.
+        if output.status.success() && output.stderr.len() == 0 {
+            return Ok(())
+        }
+
+        // Otherwise if our rustc just failed then that means that we broke the
+        // user's code with our changes. Back out everything and fall through
+        // below to recompile again.
+        if !output.status.success() {
```
[the remark this answers is not in the document]
Comment (alexcrichton) [reply cargo-fix/src/main.rs]: Definitely, I'll add a TODO

[the remark this answers is not in the document]
Comment (killercup) [reply cargo-fix/src/main.rs]: already opened #84 for that

[the remark this answers is not in the document]
Comment (alexcrichton) [reply cargo-fix/src/main.rs]: 👍

**Pull request merged**